### PR TITLE
TSM with TLS: Support Subject Alternative with IP 

### PR DESCRIPTION
**Pull request by lukas-wimmer**
Currently SAN with IP is supported only in client certificates. (Verified by the agent)

Now server side certificates can also be verified with IP using Subject Alternative Names. 

According to rfc6353, Common Name verification is deprecated. 
So I am not sure if we should also set the X509_CHECK_FLAG_NEVER_CHECK_SUBJECT flag?
... but it would change the current behavior

For more details about used flags:
https://www.openssl.org/docs/man1.1.0/man3/X509_check_host.html

### CODE REVIEW

File: snmplib/transports/snmpTLSBaseDomain.c [hunk 1]
```diff
@@ -245,88 +245,39 @@ netsnmp_tlsbase_verify_server_cert(SSL *ssl, _netsnmpTLSBaseData *tlsdata) {
...
-            int                 is_wildcarded = 0;
-            char               *compare_to;
-
-            /* see if the requested hostname has a wildcard prefix */
-            if (strncmp(tlsdata->their_hostname, "*.", 2) == 0) {
-                is_wildcarded = 1;
-                compare_to = tlsdata->their_hostname + 2;
-            } else {
-                compare_to = tlsdata->their_hostname;
+            size_t their_hostname_len = strlen(tlsdata->their_hostname);
```
Comment (bvanassche) [inline snmplib/transports/snmpTLSBaseDomain.c:248]: Please leave a blank line between declarations and statements.

File: snmplib/transports/snmpTLSBaseDomain.c [hunk 2]
```diff
@@ -245,88 +245,39 @@ netsnmp_tlsbase_verify_server_cert(SSL *ssl, _netsnmpTLSBaseData *tlsdata) {
...
-                        DEBUGMSGTL(("tls_x509:verify", "checking subjectAltname of dns:%s\n", check_name));
-                        if (strcmp(compare_to, check_name) == 0) {
-
-                            DEBUGMSGTL(("tls_x509:verify", "Successful match on a subjectAltname of dns:%s\n", check_name));
-                            return SNMPERR_SUCCESS;
-                        }
-                    }
-                }
+            for (i = 0; i < their_hostname_len; ++i) {
+                lower_hostname[i] = tolower((unsigned char) their_hostname[i]);
```
Comment (bvanassche) [inline snmplib/transports/snmpTLSBaseDomain.c:257]: The tolower() argument must be a letter. Please add an isalpha() check.

**Review by bvanassche — CHANGES_REQUESTED** [review]

File: snmplib/transports/snmpTLSBaseDomain.c (hunk 2 above)
Comment (lukas-wimmer) [reply snmplib/transports/snmpTLSBaseDomain.c:257]: As far as I have understood the function tolower() .... isaphla() is done internal. 
https://www.cplusplus.com/reference/cctype/tolower/

Anyway I will add this additional check. It won't hurt :)

File: snmplib/transports/snmpTLSBaseDomain.c (hunk 1 above)
Comment (lukas-wimmer) [reply snmplib/transports/snmpTLSBaseDomain.c:248]: Thank you for your valuable feedback, I will touch up and keep in mind for future changes as well!

**Comment by bvanassche** [comment]
I tried to build Net-SNMP with this patch applied and ran into this:
```
transports/snmpTLSBaseDomain.c: In function ‘netsnmp_tlsbase_verify_server_cert’:
transports/snmpTLSBaseDomain.c:249:13: error: ‘their_hostname’ undeclared (first use in this function); did you mean ‘their_hostname_len’?
  249 |             their_hostname = tlsdata->their_hostname;
      |             ^~~~~~~~~~~~~~
      |             their_hostname_len
transports/snmpTLSBaseDomain.c:249:13: note: each undeclared identifier is reported only once for each function it appears in
transports/snmpTLSBaseDomain.c:256:18: error: ‘i’ undeclared (first use in this function)
  256 |             for (i = 0; i < their_hostname_len; ++i) {
      |                  ^
In file included from ../include/net-snmp/output_api.h:62,
                 from ../include/net-snmp/library/snmp_client.h:32,
                 from ../include/net-snmp/varbind_api.h:102,
                 from ../include/net-snmp/library/snmp_api.h:33,
                 from ../include/net-snmp/config_api.h:90,
                 from transports/snmpTLSBaseDomain.c:36:
../include/net-snmp/library/snmp_debug.h:166:53: warning: left-hand operand of comma expression has no effect [-Wunused-value]
  166 | #define __DBGMSGTL(x)    __DBGTRACE, debugmsgtoken x, debugmsg x
      |                                                     ^
../include/net-snmp/output_api.h:68:47: note: in expansion of macro ‘__DBGMSGTL’
   68 | #define DEBUGMSGTL(x)      do {if (_DBG_IF_) {__DBGMSGTL(x);} }while(0)
      |                                               ^~~~~~~~~~
transports/snmpTLSBaseDomain.c:276:17: note: in expansion of macro ‘DEBUGMSGTL’
  276 |                 DEBUGMSGTL(("tls_x509:verify", "Successful match on a subjectAltname of IP: %s\n", their_hostname));
      |                 ^~~~~~~~~~
make[1]: *** [Makefile:100: transports/snmpTLSBaseDomain.lo] Error 1
```

**Comment by bvanassche** [comment]
I have undone the pull request.

**Pull request merged**